iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/ImportInformationSpaceJob.java

```diff
@@ -0,0 +1,404 @@
+            JavaRDD<DocumentToProject> docProjResultRelation, JavaRDD<ProjectToOrganization> projOrgResultRelation,
+            JavaRDD<IdentifierMapping> dedupResultRelation, ImportInformationSpaceJobParameters jobParams) {
+
+        // caching before calculating counts and writing on HDFS
+        docMeta.cache();
+        dataset.cache();
+        project.cache();
+        organization.cache();
+        docProjResultRelation.cache();
+        projOrgResultRelation.cache();
+        dedupResultRelation.cache();
+        
+        JavaRDD<ReportEntry> reports = generateReportEntries(sparkContext, docMeta.count(), dataset.count(),
+                project.count(), organization.count(), docProjResultRelation.count(), projOrgResultRelation.count(),
+                dedupResultRelation.count());
+
+        avroSaver.saveJavaRDD(docMeta, eu.dnetlib.iis.importer.schemas.DocumentMetadata.SCHEMA$,
+                jobParams.outputPath + '/' + jobParams.outputNameDocumentMeta);
+        avroSaver.saveJavaRDD(dataset, eu.dnetlib.iis.importer.schemas.DataSetReference.SCHEMA$,
+                jobParams.outputPath + '/' + jobParams.outputNameDatasetMeta);
+        avroSaver.saveJavaRDD(project, eu.dnetlib.iis.importer.schemas.Project.SCHEMA$,
+                jobParams.outputPath + '/' + jobParams.outputNameProject);
+        avroSaver.saveJavaRDD(organization, eu.dnetlib.iis.importer.schemas.Organization.SCHEMA$,
+                jobParams.outputPath + '/' + jobParams.outputNameOrganization);
+        avroSaver.saveJavaRDD(docProjResultRelation, DocumentToProject.SCHEMA$,
+                jobParams.outputPath + '/' + jobParams.outputNameDocumentProject);
+        avroSaver.saveJavaRDD(projOrgResultRelation, ProjectToOrganization.SCHEMA$,
+                jobParams.outputPath + '/' + jobParams.outputNameProjectOrganization);
+        avroSaver.saveJavaRDD(dedupResultRelation, IdentifierMapping.SCHEMA$,
+                jobParams.outputPath + '/' + jobParams.outputNameDedupMapping);
+
+        avroSaver.saveJavaRDD(reports, ReportEntry.SCHEMA$, jobParams.outputReportPath);
+    }
+    
+    @Parameters(separators = "=")
+    private static class ImportInformationSpaceJobParameters {
+
+        @Parameter(names = "-sparkSessionManagedOutside", required = false)
+        private Boolean sparkSessionManagedOutside;
+        
+        @Parameter(names = "-skipDeletedByInference", required = true)
+        private String skipDeletedByInference;
+        
+        @Parameter(names = "-trustLevelThreshold", required = true)
+        private String trustLevelThreshold;
+        
+        @Parameter(names = "-inferenceProvenanceBlacklist", required = true)
+        private String inferenceProvenanceBlacklist;
+        
+        @Parameter(names = "-inputRootPath", required = true)
+        private String inputRootPath;
+        
+        @Parameter(names = "-inputFormatParquet", required = true)
```

Comment:
Yes, this would be quite tempting, especially if the performance wasn't affected too much. 

On the other hand input format is some kind of a contract we are agreeing on (we, CNR, ARC) so we are going probably to stuck on 1 type in the end (parquet). Text format will be used only for testing. So maybe there is no need to introduce this JSON conversion overhead.

Reply:
OK, I've run several attempts using the original and generic approach and it seems the performance is affected quite significantly. 

Cumulative task execution time of the original solution was ~80h:
http://iis-cdh5-test-gw.ocean.icm.edu.pl:18089/history/application_1585920557248_1087/1/executors/

while using the generic approach it takes ~180h:
http://iis-cdh5-test-gw.ocean.icm.edu.pl:18089/history/application_1585920557248_1026/1/executors/

I have repeated both versions several times and obtained repeatable results.

The change I am going to introduce is replacing `inputFormatParquet` boolean flag I didn't like anyway with `inputFormat` and for now allowing two possible values: `json` and `paquet`. Alternatively I could allow an arbitrary value and fallback to generic approach but since it turned out to have performance overhead I would probably prefer to stick to the mentioned two. I guess it will be refactored anyway once we agree with Claudio for a single allowed format.

Reply:
Yes, I think you're right to allow for two input format at this time. I didn't think that this roundtrip to json and back will be so costly in terms of performance.